### ISSUE TITLE
add dependencies

### ISSUE THREAD

**Issue by andreihaidu** [code-iai/semrec#7]
maybe updating the README.md file with required dependencies

I had to install `libncurses5-dev` for one error.

And currently I am getting something  with 
```
semrec/3rdparty/json-c/json_inttypes.h:5:25: fatal error: json_config.h: No such file or directory
 #include "json_config.h"
```

**Comment by fairlight1337** [comment]
Hey,

I remember us fixing this after talking about that in person. More specifically, the following dependencies were missing:

  * `libncurses5-dev`
  * `automake`
  * `autoconf`

I addressed these in the `README.md` (see 62ca4dfd4fabcf1af3f2d4a13bca272482436e47). ncurses is needed for building the `console` plugin, which I think could be make optional. But for now, installing it shouldn't be that much of a burden.

The latter two are needed for compiling (or, more precisely, configuring) the version of `json-c`, that is checkec out during build automatically. These must be installed, as `json-c` is an integral part of `semrec`. I'll see after adding a check for those that will output a more meaningful error message.

Thanks for the hints. Did you come across any other missing dependencies?

**Comment by andreihaidu** [comment]
Works wonderfull! Thanks! :)

**Comment by airballking** [comment]
Hi guys,

I am getting the same error that @andreihaidu described even though I installed ```libncurses5-dev```, ```automake```, and ```autoconf```. Are you sure that installing those packages fixes the problem with the missing ```json_config.h```?

**Comment by fairlight1337** [comment]
There is a workaround for that issue. Inside 3rdparty/json-c (or similar), find "build_flag" and delete it. Then rebuild.

I'll look into something more sophisticated later.

Cheers

**Comment by airballking** [comment]
Thanks for the fix: That solved my problem!

**Comment by airballking** [comment]
Ping because I ran into this issue again when installing semrec on the PC at DLR.

**Comment by airballking** [comment]
Just out of curiosity: What exactly are autogen and autoconf doing for 3rdparty/json-c? Do they offer any flexibility we need?

**Comment by fairlight1337** [comment]
Nope, 0.12 has a feature I use in the code, which is not yet present downstream.

`automake` and `autoconf` are the build tools of choice for `libjson-c`. There's not much changeable about that; but it should be possible to add them as system dependencies, no?

I didn't get around to properly create the respective rosdep files for that, yet.

**Comment by airballking** [comment]
To bad that 0.11 is not enough for us. I was hoping to use ```libjson-c``` as system-dependency. Anyway. Have to wait for Ubuntu 16.04, then.

By the way, I already added ```automake``` and ```autoconf``` to the rosdep files, and Tully merged last week. So, one can use them as system deps, now.

**Comment by fairlight1337** [comment]
Okay, I had a closer look.

It seems as if the reason for checking out the git version of `libjson-c` was grounded in a prior version shipped with Ubuntu 12.04. That wasn't too obvious, as the 3rdparty scripts in `semrec` checkout the *newest* available version, not a specific one (baddy on my side there).

I just recompiled locally without having the 3rdparty version of `libjson-c` checked out (aka purged the whole 3rdparty folder) and it works (even some deprecation warnings went away, yay!).

One more reason not to push that change just yet is the fact that the Travis CI build farm is still configured to use Ubuntu 12.04. Recently, they made 14.04 available for build testing. I'll update the build script and push the changes. The `libjson-c-dev` and `libjson-c2` can then be used from the system deps.

I'm re-opening the issue until things are settled. Thanks for keeping this alive, @airballking.

**Comment by fairlight1337** [comment]
I pushed it. Let's hope this works; I'll keep the ball rolling once the build completes/fails. After that, the rosdeps need to be updated for the package.

**Comment by fairlight1337** [comment]
Okay. No hydro on 14.04, right. I'm merging the bulid script to indigo, that's gonna take a couple moments.

**Comment by airballking** [comment]
That we can use ```libjson-c``` as a system dep on Ubuntu 14.04 is great news! Looking forward to the merge. :)

**Comment by fairlight1337** [comment]
Alright, I just verified that the Travis CI build is back to normal. The `3rdparty` directory is gone (it only held `json-c`) and `libjson-c` is now used from the debs. I updated the `package.xml` accordingly (only `build_depend`s yet, though), so `rosdep` should be happy with that as well.

@airballking, can you give me a heads up whether that works for you?

The whole thing now doesn't compile under 12.04 anymore, but as 14.04 + Indigo is LTS anyway, that should be a safe bet.

**Comment by airballking** [comment]
After manually installing the two system libs, it compiled. :)

However, the rosdep rules for both are not yet in place. Do you plan to add them?

**Comment by fairlight1337** [comment]
Great!

Would love to add the rosdep rules; do you do that manually or is there an automated process (in the flavor of bloom) for that?

**Comment by airballking** [comment]
To add rosdep rules, you need to fork the rosdistro repository, manually add the new rules, check them, and then send a pull request for OSRF to merge. The process is documented here:

http://docs.ros.org/independent/api/rosdep/html/contributing_rules.html

**Comment by fairlight1337** [comment]
Thanks, great link.

During the process of doing that, I realized that the already existing entry `libjson0-dev` covers the necessary packages. I adapted the `package.xml` in `semrec` accordingly. https://github.com/code-iai/semrec/commit/f1e8a6d3ef9150bd64ccdc659833b643d7f75cd3 should fix that.

If you can verify that it works as expected, we can finally close the issue I guess.

**Comment by airballking** [comment]
I compiled ```semrec``` from source on a new PC at DLR, today. That worked OK. So, I guess removing the 3rd-party stuff worked out fine. :)

On a side-note, I had to install libconfig++ through apt-get and it is not listed in the package.xml. Do you wanna change that?

**Comment by fairlight1337** [comment]
That's probably a good idea. Is it referenceable through rosdep already?

**Comment by airballking** [comment]
I don't know.

**Comment by fairlight1337** [comment]
I'll check later and add it to the packages.xml if it's the case. But I guess it is.

**Comment by fairlight1337** [comment]
@bbferka I remember talking to you about this issue; do you still have installation issues with the dependencies? If yes, which ones are missing from the `package.xml`?

**Comment by bbferka** [comment]
I did not use rosdep install before compiling and rolibjson-c-dev and libconfig++-dev where the two libs missing for me. Another issue is during compilation the dessignator_msgs don't get generated in time, so first call to catkin_make fails.  This can be solved by adding the dependency to the library explicitly,  e.g. for this lib:

add_dependencies(sr_plugin_symboliclog designator_integration_msgs_gencpp)
